### PR TITLE
Add `--machine` support for `flutter attach`

### LINKED ISSUE #19045: Add a --machine option for `flutter attach`
@devoncarew @mit-mit Looks like `attach` doesn't currently support `--machine` (it also gives a weird error: `The --machine flag is only valid with the --version flag`).

It's currently implemented as its own command and creates a `HotRunner`. Our JSON support is all inside Daemon. Does it makes sense for AttachCommand to work more like RunCommand and spawn a daemon internally (so it gets all the AppDomain stuff)? (I'm not sure if we're going to end up piling lots of stuff in Daemon; but this does want to reuse a lot of the run functionality).

### PR DESCRIPTION
**Pull request by DanTup**
I've split some of the code out of `startApp` in `Daemon` into a new method (`launch`) so that it can be reused by `attach` (this handles sending app.start messages, etc.) and changed `AttachCommand` to create a `Daemon` when run in `--machine`  mode (similar to how `RunCommand` does for `flutter run --machine`).

I also bumped the daemon protocol version so that clients can detect whether `flutter attach --machine` is supported (though note, you'll need to have spawned some other type of daemon in order to have received the protocol version - probably this is no problem because editors will spawn the device daemon well before the user tries to attach).

Tests are still outstanding; but I wanted to get feedback (and allow this to be tried out - I have a working implementation in Dart Code I'll make a preview build of soon). I'd like to build the tests using the new integration test driver I was working on (#18865 and #18878) so that they can be run on all builds/PRs instead of just devicelab.

Fixes #19045.

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/attach.dart
Comment (mravn-google) [inline packages/flutter_tools/lib/src/commands/attach.dart]: Trailing commas would help improve formatting here and several places below, cf https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#if-you-have-a-newline-after-some-opening-punctuation-match-it-on-the-closing-punctuation

File: packages/flutter_tools/lib/src/commands/daemon.dart
Comment (mravn-google) [inline packages/flutter_tools/lib/src/commands/daemon.dart]: Perhaps use a `typedef` to give a name and return type to this function. The declaration here does not provide any hint that `run` should be asynchronous.

It might also aid readability to rename `run`. I'm not sure what it is supposed to do, or why we have both a `runner` and a `run` parameter.

File: packages/flutter_tools/lib/src/commands/daemon.dart
Comment (mravn-google) [inline packages/flutter_tools/lib/src/commands/daemon.dart]: Can we use a `Completer<void>` and `.then<void>((_) { ... })`?

**Review by mravn-google — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/commands/daemon.dart
Comment (DanTup) [reply packages/flutter_tools/lib/src/commands/daemon.dart]: We need the runner inside here to create the AppInstance but we also need to kick off either starting the app or attaching to it (depending on where it's called from). They have different params (`run` needs a `route`) so this was the best I could come up with.

I've added a typedef and renamed to `runOrAttach`; but definitely happy to change this if you have better ideas :)

**Comment by DanTup** [comment]
(Holding off landing this until it has some integration tests; waiting on #19085 to make that easier).

**Comment by DanTup** [comment]
@devoncarew @scheglov I'm having problems getting tests working on flutter-tester for this. If I run `flutter attach` with a simulator open, then run `flutter run` in another terminal, everything works fine (the app runs from `flutter run` and then `flutter attach` finds it). However, on `flutter-tester` the attach process never finds it. I don't know if this is due to how `flutter-tester` works (it's spawned inside `flutter run`) or wether this should work (eg. when `flutter-tester` is running inside `flutter run`, should it be visible to other flutter process like other devices would be?)?

**Comment by devoncarew** [comment]
As discussed off-line, this could possible be solved by passing an explicit post to `flutter attach`. Another thing to try is to have a flag which would cause `flutter-tester` to be listed as a device (`flutter attach --list-tester-device`).

**Comment by DanTup** [comment]
I tried passing the port and it crashed because the `portForwarder` on the `FlutterTester` device is `null`. Possibly we could make a fake port forwarder that doesn't forward, but just returns the same port for FlutterTester, but... I think we'll see other issues - like this code calls `device.getLogReader()` but for FlutterTester I think this is an in-process string stream; so the attach process will likely have its own (and not receive output from the test device running in the other process)?

I guess an important question to answer is whether the test device should be visible outside of its own process (besides over observatory). Should we be able to connect to it (eg. and read logs) like we can a real device or simulator? If so, we may be missing some implementation. If we shouldn't, then we might not be able use it for testing this way.

(cc @scheglov - I think you worked on this, so may have some thoughts?).

**Comment by scheglov** [comment]
FlutterTester is almost like a real device, same VM, same Observatory, it is just executed not in simulator, but in a more lightweight process. Not sure about which logs we are talking. There are no Android device logs, because there is no Android, but there should be Flutter compilation logs, and probably log entries from the Flutter app. In theory. Whether any of this is implemented, that's another question :-)

**Comment by DanTup** [comment]
What I'd like to be able to do is open two terminals and run the following:

- `flutter run -d flutter-tester`
- `flutter attach -d flutter-tester --debug-port x`

This works fine with a real device (even without `--debug-port x`) but fails on `flutter-tester`. I believe attach needs a few things to work:

1. A connection to observatory
2. A PortForwarder
3. A log reader

**1** is easy, **2** I think we can just make a fake port forwarder for the tester (the device is the local machine, so it's just a no-op forwarder), however **3** I'm not sure about. I believe if the user has `print` statements in their app (for example), they would come via the log reader?

I don't know if there's a mechanism that would allow us to get the output from a `flutter-tester` process that we don't have access to stdout/stderr for (eg. it's spawned in the `flutter run` process and we want it inside `flutter attach`).

Possibly for now we could just make it not get the log output - I don't need it for these specific tests; I was more interested in how similar to a real device we want it to be.

So how about for now:

1. We make a no-op PortForwarder for the FlutterTester device (so it just returns the same port you asked for)
2. There will be no output captured from flutter-tester devices when connected remotely

(I don't think that'd be much effort, so I'll take a shot at it later in the absence of objections)

**Comment by DanTup** [comment]
Ok, I've pushed a basic test for this - it includes a no-op port forwarder in the test device and we just won't get logs for now.

(@mravn-google apologies - I probably should've put it in a new PR to make it easier for you to just review the new code - it may be easier to review by-changeset from the ones from last 2 days).

The test is very basic - I have a note to come back to this (and the other reload tests) to make them more complete (eg. make sure a reload actually happens and code is re-executed, rather than just that the daemon said it reloaded).

File: packages/flutter_tools/lib/src/tester/flutter_tester.dart
Comment (devoncarew) [inline packages/flutter_tools/lib/src/tester/flutter_tester.dart]: `// TODO(dantup):` in order to give people the best idea of who to contact for more information

File: packages/flutter_tools/lib/src/commands/daemon.dart
Comment (devoncarew) [inline packages/flutter_tools/lib/src/commands/daemon.dart]: formatting looks a little wonky here - perhaps have the `({` on the previous line?

**Review by devoncarew — APPROVED** [review]
test lgtm

File: packages/flutter_tools/lib/src/tester/flutter_tester.dart
Comment (DanTup) [reply packages/flutter_tools/lib/src/tester/flutter_tester.dart]: Oops, this was stale - this is implemented.

File: packages/flutter_tools/lib/src/commands/daemon.dart
Comment (DanTup) [reply packages/flutter_tools/lib/src/commands/daemon.dart]: I think I've improved it... dartfmt did a bad job with it and I couldn't make it look nice. Hopefully it's better now.

Comment (DanTup) [reply packages/flutter_tools/lib/src/commands/daemon.dart]: dartfmt does this even with that:

screen shot 2018-07-12 at 5 33 39 pm

Here's how it looks as committed now:

screen shot 2018-07-12 at 5 34 18 pm

I considered this but it's *way* over 80 chars!

screen shot 2018-07-12 at 5 35 52 pm

**Comment by DanTup** [comment]
@devoncarew When merged with `master` some tests started failing like this:

```
type '_AsyncCompleter<void>' is not a subtype of type 'Completer<Null>'","trace":"
#0      AppDomain.startApp.<anonymous closure> (package:flutter_tools/src/commands/daemon.dart:383:38)
#1      AppDomain.launch.<anonymous closure> (package:flutter_tools/src/commands/daemon.dart:432:26)
<asynchronous suspension>
#2      AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:142:29)
<asynchronous suspension>
#3      _rootRun (dart:async/zone.dart:1126:13)
#4      _CustomZone.run (dart:async/zone.dart:1023:19)
#5      _runZoned (dart:async/zone.dart:1518:10)
#6      runZoned (dart:async/zone.dart:1465:12)
#7      AppContext.run (package:flutter_tools/src/base/context.dart:141:18)
<asynchronous suspension>
#8      AppInstance._runInZone (package:flutter_tools/src/commands/daemon.dart:781:20)
#9      AppDomain.launch (package:flutter_tools/src/commands/daemon.dart:430:1...
```

I'd changed some `Completer<Null>` to `Completer<void>` in here (which seemed sensible, I presume the `Null` was because we previously didn't support `void` as a generic type arg?) but looks like they didn't match everywhere. I had to change it in a few more places to fix this, but I'm slightly nervous that it's further outside the code I modified. 

Can you check e198d00ffdf0a98769bb09c737c6b4746a319571 to see if these changes seem safe to you? I'm nervous about things that can blow up at runtime with no compile time warnings.

**Comment by devoncarew** [comment]
Yes, with Dart 2 we'll need to ensure we have good test coverage, or switch to using additional analysis settings, like `implicit-casts: false`.

**Comment by DanTup** [comment]
> Yes, with Dart 2 we'll need to ensure we have good test coverage

I wasn't so concerned about the new code here (it's new, so it's not going to break anyone) but those last changes affected run/runner which are pretty important.

I'll have a go at disabling implicit casts later and scan through (though there might be a lot of noise!) before I land this to catch anything obvious. Maybe between our own integration tests, device lab, etc., there's reasonable coverage.

Just to confirm though - is changing `Completer<Null>/Future<Null>` to use `void` a sensible thing to do? Was `Null` a workaround for not being able to put `void` here before?

**Comment by DanTup** [comment]
> or switch to using additional analysis settings, like `implicit-casts: false`.

Couldn't get this to work... I found [this issue](https://github.com/dart-lang/sdk/issues/32235) but even using `--no-implicit-casts` I couldn't get warnings even on code I know had them!

The tests are all passing, so I'm probably overly-paranoid. I'll land it when it's quiet (build is red now, so in the morning if it's fixed) and keep an eye on the devicelabs just in case.

**Pull request merged**
